### PR TITLE
v1.0 backports 2018-08-03

### PR DESCRIPTION
**Pull request by joestringer**
* #4492
  * Conflicts again @ianvernon, details below
  * See also the backport approach taken in #5081
* #5074
  * Depended on #4286 , so I backported this to resolve the conflict.

Conflict for  ab77b79b4c38 ("daemon: refactor updating of CNP Node Status"):

```
diff --cc daemon/k8s_watcher.go
index 59d966bcd6cd,942f2d4b6ee5..000000000000
--- a/daemon/k8s_watcher.go
+++ b/daemon/k8s_watcher.go
@@@ -1452,40 -1477,39 +1471,72 @@@ func (d *Daemon) addCiliumNetworkPolicy
                                                Warn("Error parsing new CiliumNetworkPolicy rule")
                                }
  
++<<<<<<< HEAD
 +                              cnpns := cilium_v2.CiliumNetworkPolicyNodeStatus{
 +                                      OK: true,
 +                                      // If the deadline was reached then not all endpoints are enforcing
 +                                      // the given policy.
 +                                      Enforcing:   waitForEPsErr == nil,
 +                                      Revision:    rev,
 +                                      LastUpdated: cilium_v2.NewTimestamp(),
 +                              }
 +
 +                              // Most important is the error while adding the CNP
 +                              if err != nil {
 +                                      cnpns = cilium_v2.CiliumNetworkPolicyNodeStatus{
 +                                              Error:       err.Error(),
 +                                              OK:          false,
 +                                              LastUpdated: cilium_v2.NewTimestamp(),
 +                                      }
 +                              } else if err2 != nil {
 +                                      cnpns = cilium_v2.CiliumNetworkPolicyNodeStatus{
 +                                              Error:       err2.Error(),
 +                                              OK:          false,
 +                                              LastUpdated: cilium_v2.NewTimestamp(),
 +                                      }
 +                              }
 +
 +                              nodeName := node.GetName()
 +                              serverRuleCpy.SetPolicyStatus(nodeName, cnpns)
 +                              ns := k8sUtils.ExtractNamespace(&serverRuleCpy.ObjectMeta)
 +
 +                              _, err2 = ciliumNPClient.CiliumV2().CiliumNetworkPolicies(ns).Update(serverRuleCpy)
 +                              if err2 == nil {
++=======
+                               var err3 error
+ 
+                               // Update the status of whether the rule is enforced on this node.
+                               // If we are unable to parse the CNP retrieved from the store,
+                               // or if endpoints did not reach the desired policy revision
+                               // after 30 seconds, then mark the rule as not being enforced.
+                               if policyImportErr != nil {
+                                       // OK is false here because the policy wasn't imported into
+                                       // cilium on this node; since it wasn't imported, it also
+                                       // isn't enforced.
+                                       err3 = updateCNPNodeStatus(serverRuleCpy, false, false, policyImportErr, rev, cnp.Annotations)
+                               } else if ruleCopyParseErr != nil {
+                                       // This handles the case where the initial instance of this
+                                       // rule was imported into the policy repository successfully
+                                       // (policyImportErr == nil), but, the rule has been updated
+                                       // in the store soon after, and is now invalid. As such,
+                                       // the rule is not OK because it cannot be imported due
+                                       // to parsing errors, and cannot be enforced because it is
+                                       // not OK.
+                                       err3 = updateCNPNodeStatus(serverRuleCpy, false, false, ruleCopyParseErr, rev, cnp.Annotations)
+                               } else {
+                                       // If the deadline by the above context, then not all
+                                       // endpoints are enforcing the given policy, and
+                                       // waitForEpsErr will be non-nil.
+                                       err3 = updateCNPNodeStatus(serverRuleCpy, waitForEPsErr == nil, true, waitForEPsErr, rev, cnp.Annotations)
+                               }
+ 
+                               if err3 == nil {
++>>>>>>> daemon: refactor updating of CNP Node Status
                                        scopedLog.WithField("status", serverRuleCpy.Status).Debug("successfully updated with status")
                                } else {
-                                       return err2
+                                       return err3
                                }
+ 
                                return waitForEPsErr
                        },
                },
@@@ -1550,10 -1675,13 +1702,17 @@@ func (d *Daemon) updateCiliumNetworkPol
                logfields.K8sAPIVersion:                    oldRuleCpy.TypeMeta.APIVersion,
                logfields.CiliumNetworkPolicyName + ".old": oldRuleCpy.ObjectMeta.Name,
                logfields.K8sNamespace + ".old":            oldRuleCpy.ObjectMeta.Namespace,
++<<<<<<< HEAD
 +              logfields.CiliumNetworkPolicyName + ".new": newRuleCpy.ObjectMeta.Name,
 +              logfields.K8sNamespace + ".new":            newRuleCpy.ObjectMeta.Namespace,
++=======
+               logfields.CiliumNetworkPolicyName:          newRuleCpy.ObjectMeta.Name,
+               logfields.K8sNamespace:                     newRuleCpy.ObjectMeta.Namespace,
+               "annotations.old":                          oldRuleCpy.ObjectMeta.Annotations,
+               "annotations":                              newRuleCpy.ObjectMeta.Annotations,
++>>>>>>> daemon: refactor updating of CNP Node Status
        }).Debug("Modified CiliumNetworkPolicy")
  
 -      d.deleteCiliumNetworkPolicyV2(oldRuleCpy)
        d.addCiliumNetworkPolicyV2(ciliumV2Store, newRuleCpy)
  }
 
```

Applied this additional diff to fix merge conflicts:

```
diff --cc daemon/k8s_watcher.go
index 59d966bcd6cd,5583fb40f57a..000000000000
--- a/daemon/k8s_watcher.go
+++ b/daemon/k8s_watcher.go
@@ -1441,12 +1442,7 @@ func updateCNPNodeStatus(cnp *cilium_v2.CiliumNetworkPolicy, enforcing, ok bool,
        cnp.SetPolicyStatus(nodeName, cnpns)
        ns := k8sUtils.ExtractNamespace(&cnp.ObjectMeta)
 
-       switch {
-       case ciliumUpdateStatusVerConstr.Check(k8sServerVer):
-               _, err2 = ciliumNPClient.CiliumV2().CiliumNetworkPolicies(ns).UpdateStatus(cnp)
-       default:
-               _, err2 = ciliumNPClient.CiliumV2().CiliumNetworkPolicies(ns).Update(cnp)
-       }
+       _, err2 = ciliumNPClient.CiliumV2().CiliumNetworkPolicies(ns).Update(cnp)
        return err2
 }
 
@@@ -1492,6 -1414,38 +1510,36 @@@ func (d *Daemon) addCiliumNetworkPolicy
        )
  }

+ func updateCNPNodeStatus(cnp *cilium_v2.CiliumNetworkPolicy, enforcing, ok bool, err error, rev uint64, annotations map[string]string) error {
+       var (
+               cnpns cilium_v2.CiliumNetworkPolicyNodeStatus
+               err2  error
+       )
+ 
+       if err != nil {
+               cnpns = cilium_v2.CiliumNetworkPolicyNodeStatus{
+                       Enforcing:   enforcing,
+                       Error:       err.Error(),
+                       OK:          ok,
+                       LastUpdated: cilium_v2.NewTimestamp(),
 -                      Annotations: annotations,
+               }
+       } else {
+               cnpns = cilium_v2.CiliumNetworkPolicyNodeStatus{
+                       Enforcing:   enforcing,
+                       Revision:    rev,
+                       OK:          ok,
+                       LastUpdated: cilium_v2.NewTimestamp(),
 -                      Annotations: annotations,
+               }
+       }
+ 
+       nodeName := node.GetName()
+       cnp.SetPolicyStatus(nodeName, cnpns)
+       ns := k8sUtils.ExtractNamespace(&cnp.ObjectMeta)
+ 
+       _, err2 = ciliumNPClient.CiliumV2().CiliumNetworkPolicies(ns).Update(cnp)
+       return err2
+ }
+ 
  func (d *Daemon) deleteCiliumNetworkPolicyV2(cnp *cilium_v2.CiliumNetworkPolicy) {
        scopedLog := log.WithFields(logrus.Fields{
                logfields.CiliumNetworkPolicyName: cnp.ObjectMeta.Name,
@@@ -1550,8 -1568,10 +1662,8 @@@ func (d *Daemon) updateCiliumNetworkPol
                logfields.K8sAPIVersion:                    oldRuleCpy.TypeMeta.APIVersion,
                logfields.CiliumNetworkPolicyName + ".old": oldRuleCpy.ObjectMeta.Name,
                logfields.K8sNamespace + ".old":            oldRuleCpy.ObjectMeta.Namespace,
-               logfields.CiliumNetworkPolicyName + ".new": newRuleCpy.ObjectMeta.Name,
-               logfields.K8sNamespace + ".new":            newRuleCpy.ObjectMeta.Namespace,
+               logfields.CiliumNetworkPolicyName:          newRuleCpy.ObjectMeta.Name,
+               logfields.K8sNamespace:                     newRuleCpy.ObjectMeta.Namespace,
 -              "annotations.old":                          oldRuleCpy.ObjectMeta.Annotations,
 -              "annotations":                              newRuleCpy.ObjectMeta.Annotations,
        }).Debug("Modified CiliumNetworkPolicy")

        d.addCiliumNetworkPolicyV3(ciliumV2Store, newRuleCpy)
```

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/5091)

### CODE REVIEW

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Added upgrade-impact label due to backport of #4286

**Comment by ianvernon** [comment]
Thanks for posting the diffs @joestringer !!

**Review by ianvernon — APPROVED** [review]

**Comment by ianvernon** [comment]
test-missed-k8s

**Comment by ianvernon** [comment]
test-upstream-k8s

**Comment by ianvernon** [comment]
test-docs-please

**Comment by ianvernon** [comment]
Dang, looks like we don't have upstream k8s tests in v1.0. Ignore that build failure.

**Comment by joestringer** [comment]
Missed-k8s failed due to some kind of connectivity issue on startup:
```
    k8s1-1.8: Step 7/15 : FROM quay.io/cilium/cilium-runtime:2018-06-04
    k8s1-1.8: Get https://quay.io/v2/cilium/cilium-runtime/manifests/2018-06-04: Get https://quay.io/v2/auth?scope=repository%3Acilium%2Fcilium-runtime%3Apull&service=quay.io: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
```

Retrying.

**Comment by joestringer** [comment]
test-missed-k8s

**Comment by joestringer** [comment]
Can't seem to run missed k8s tests. Not sure how it got stuck with a failed run, but that's the one which failed on bootup of the machine.

**Comment by eloycoto** [comment]
test-missed-k8s

**Review by tgraf — CHANGES_REQUESTED** [review]
1.0 backport and upgrade-impact is almost a contradiction on its own. I don't feel safe backporting a large change such as BPF map root and CNP status refactoring to 1.0 at this point.

Why are these changes crucial? Do we have users with issues without these fixes?

**Comment by joestringer** [comment]
I believe that the node status piece (#4492) was intended to fix the correct representation of the "enforcing" flag. Perhaps @ianvernon has some more context on this. It was originally a community report (#4248), however at a glance it looks to me like a feature so it seems very reasonable to avoid backporting to 1.0.

The mount path change (#4286) for parameter to pass to the monitor I pulled in as a dependency to resolve the conflict for the CPU usage issue with cilium-monitor (#5074).  I can attempt to backport again without pulling in this dependency. We should be able to avoid this upgrade-impact that way.

Does that sound reasonable, @tgraf?

**Comment by raybejjani** [comment]
@joestringer On pulling in #4286 for #5074. I doubt 1.0 received the bpffs mount changes so it should be ok to not include it, and doing that should be straight-forward. #4286 added a `bpfRoot` parameter that can be undone and read as before #4286, or simply use the hardcoded value from before #4286 as the parameter (as opposed to using the more dynamic `getRoot()` call). I can give it a go if you prefer (and since I'm claiming it's easy ;) )

More generally, though, why do we think this has an upgrade impact? This code controls spawning cilium-node-monitor. It is strictly a child of the currently running cilium-agent. In all cases the version of cilium-node-monitor would match that of cilium-agent. In particular #4286 is relatively agnostic to whether where the mount is, beyond making it a parameter. In any case, the default for cilium-node-monitor is the hardcoded value from before, so an older cilium-agent should still work out ok.

tl;dr I don't think there is an upgrade impact

**Comment by joestringer** [comment]
@raybejjani thanks for chiming in. I didn't think deeply through the upgrade impact, it's just that at a cursory glance it seemed like commandline arguments were changing and this can sometimes have unforeseen impact (particularly eg with cilium-agent needing to sync with the k8s YAMLs). Your explanation for why it does not impact upgrade makes sense.

Part of the reason I put the `upgrade-impact` label on was to signal that I'm not confident I understand the impact of the changes. From that perspective I think it's managed to draw enough attention.

**Comment by joestringer** [comment]
I've re-attempted this backport, scaling down the changes. See #5119 for more details. Closing.

**Comment by raybejjani** [comment]
> Your explanation for why it does not impact upgrade makes sense.

Full disclosure: I didn't test my reasoning, so I can if you'd like ;) (in particular, I don't recall whether that bpf path has changed at some other point, rendering my reasoning false).
